syntax = "proto3";
package grpcdemo;


// ================================================================================
// DEMO MESSAGES for API Eigen Example
// ================================================================================

// We define the data-types involved in the data transfer. Bear in mind that the amount
// of space in memory that an Integer and a Double occupy is different. Thus it is
// important to know how to breakdown our chunk of bytes.
enum DataType{
    INTEGER = 0;
    DOUBLE = 1;
  }

// Interface definition for our Vector
message Vector {
    DataType data_type = 1;
    int32 vector_size = 2;
    bytes vector_as_chunk = 3;
  }

// Interface definition for our Matrix
message Matrix {
    DataType data_type = 1;
    int32 matrix_rows = 2;
    int32 matrix_cols = 3;
    bytes matrix_as_chunk = 4;
}

// ================================================================================
// DEMO SERVICE for API Eigen Example
// ================================================================================

// Interface exported by the server. Different to REST... we do not require to post
// (i.e. store) in the end server the Objects of the interface (e.g. Vector, Matrix)
service GRPCDemo {
    // Vector-related methods
    rpc AddVectors(stream Vector) returns (Vector) {}
    rpc MultiplyVectors(stream Vector) returns (Vector) {}

    // Matrix-related methods
    rpc AddMatrices(stream Matrix) returns (Matrix) {}
    rpc MultiplyMatrices(stream Matrix) returns (Matrix) {}
}